### PR TITLE
♻️Reaction API change

### PR DESCRIPTION
**Pull request by mszylkowski**
Updated API to design doc specs:
```
GET http://localhost:3000/reactions/<canonicalURL>+<pageId>+<id>?reactionType=0&clientId=amp-281391283
POST http://localhost:3000/reactions/<canonicalURL>+<pageId>+<id>/vote?reactionType=0&clientId=amp-281391283 {"optionSelected": 2}

response: {
  "options": [
    {"totalCount": 4, "selectedByUser": true,  "optionIndex": 0},
    {"totalCount": 6, "selectedByUser": false, "optionIndex": 1},
    {"totalCount": 2, "selectedByUser": false, "optionIndex": 2},
    {"totalCount": 1, "selectedByUser": false, "optionIndex": 3}
  ]
}
```

- Removed hasUserResponded and totalResponsesCount from the server since the backend could retrieve data for options that were removed in the frontend, and that would lead to percentages not adding up to 100%. Calculating the percentages only for the visible options guarantees to follow the percentages calculations pipeline with only the percentages that will be visible.
- Backend API was updated to reflect latest changes in the design doc, following REST principles.
- Added tests for reaction.

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story-reaction-quiz.css [hunk 1]
```diff
@@ -267,11 +267,11 @@
      left: 101% !important;
  }
  
- .i-amphtml-story-reaction-post-selection.i-amphtml-story-reaction-has-data .i-amphtml-story-reaction-option-selected[correct].i-amphtml-story-reaction-quiz-option {
+ .i-amphtml-story-reaction-post-selection .i-amphtml-story-reaction-option-selected[correct].i-amphtml-story-reaction-quiz-option {
      background: var(--correct-color) !important;
  }
  
- .i-amphtml-story-reaction-post-selection.i-amphtml-story-reaction-has-data .i-amphtml-story-reaction-option-selected:not([correct]).i-amphtml-story-reaction-quiz-option {
+ .i-amphtml-story-reaction-post-selection .i-amphtml-story-reaction-option-selected:not([correct]).i-amphtml-story-reaction-quiz-option {
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction-quiz.css:274]: Quiz was not showing the correct background of the selected option when there was no backend, with this change it was fixed.

File: extensions/amp-story/1.0/amp-story-reaction-quiz.js [hunk 2]
```diff
@@ -103,8 +105,8 @@ export class AmpStoryReactionQuiz extends AmpStoryReaction {
 
     // First child must be heading h1-h3
     if (!['h1', 'h2', 'h3'].includes(promptInput.tagName.toLowerCase())) {
-      dev().error(
-        TAG,
+      devAssert(
+        false,
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction-quiz.js:109]: dev().error() didn't get caught on the tests so the tests weren't working at all, devAssert does throw the error so tests work now.

File: extensions/amp-story/1.0/test/test-amp-story-reaction-quiz.js [hunk 3]
```diff
@@ -195,38 +126,58 @@ describes.realWin(
...
     it('should throw an error with fewer than one prompt', () => {
       populateQuiz(win, ampStoryQuiz.element, 0);
-      expect(ampStoryQuiz.buildCallback).to.throw();
+      allowConsoleError(() => {
+        expect(() => {
+          ampStoryQuiz.buildCallback();
+        }).to.throw(
+          /The first child must be a heading element <h1>, <h2>, or <h3>/
+        );
+      });
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/test/test-amp-story-reaction-quiz.js:135]: This is how errors can get caught on the tests, now working

File: extensions/amp-story/1.0/test/test-amp-story-reaction-quiz.js [hunk 4]
```diff
@@ -195,38 +126,58 @@ describes.realWin(
...
+      allowConsoleError(() => {
+        expect(() => {
+          ampStoryQuiz.buildCallback();
+        }).to.throw(/Improper number of options/);
+      });
+    });
+
+    it('should not throw an error with three options and one prompt', () => {
+      populateQuiz(win, ampStoryQuiz.element, 1, 3);
+      expect(() => ampStoryQuiz.buildCallback()).to.not.throw();
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/test/test-amp-story-reaction-quiz.js:158]: Added test that checks for valid configuration

File: extensions/amp-story/1.0/test/test-amp-story-reaction.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
+/**
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/test/test-amp-story-reaction.js:1]: Most of the tests are moved from the previous quiz test file

File: src/url.js [hunk 6]
```diff
@@ -650,3 +650,19 @@ export function checkCorsUrl(url) {
...
+ */
+export function appendPathToUrl(url, path) {
+  const mainAndFragment = url.split('#', 2);
+  const mainAndQuery = mainAndFragment[0].split('?', 2);
+  let newUrl =
+    mainAndQuery[0].replace(/\/$/, '') + '/' + path.replace(/^\//, '');
+  newUrl += mainAndQuery[1] ? '?' + mainAndQuery[1] : '';
+  newUrl += mainAndFragment[1] ? '#' + mainAndFragment[1] : '';
+  return newUrl;
+}
```
Comment (mszylkowski) [inline src/url.js:668]: Added function that adds a path to a url before the query params and fragment. Check tests on test-url.js

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 7]
```diff
@@ -431,72 +438,52 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
-  updateReactionData_(reactionValue) {
-    this.executeReactionRequest_(
-      dict({
-        'method': 'POST',
-      }),
-      reactionValue
-    ).catch((error) => {
-      dev().error(TAG, error);
+    return this.executeReactionRequest_('GET').then((response) => {
+      this.handleSuccessfulDataRetrieval_(response);
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction.js:444]: Catch was moved to executeReactionRequest

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 8]
```diff
@@ -431,72 +438,52 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
+    if (!this.reactionId_) {
       const pageId = closest(dev().assertElement(this.element), (el) => {
         return el.tagName.toLowerCase() === 'amp-story-page';
       }).getAttribute('id');
-
-      this.reactionId_ = `CANONICAL_URL#page=${pageId}`;
+      this.reactionId_ = `CANONICAL_URL+${pageId}`;
+      if (this.element.hasAttribute('id')) {
+        this.reactionId_ += '+' + this.element.getAttribute('id');
+      }
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction.js:469]: Element id is a better tracker than page ID in any case, so we definitely should use the element id. Removing the page ID would also help if someone wanted to move a quiz from one page to another, but we need to ensure every reaction element has an ID before going that route. But still, adding the id allows for multiple reactions to live in the same page without errors until we can validate that there must only be one per page (if we want).

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 9]
```diff
@@ -520,47 +505,37 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
-    }
+    devAssert(
+      response && 'options' in response,
+      `Invalid reaction response, expected { data: ReactionResponseType, ...} but received ${response}`
+    );
+    const numOptions = this.rootEl_.querySelectorAll(
+      '.i-amphtml-story-reaction-option'
+    ).length;
+    this.updateReactionOnDataRetrieval_(
+      response['options'].splice(0, numOptions)
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction.js:518]: Only keep the first N options (N= options on the frontend quiz). This ensures the percentages add up to 100 if the backend returns unused options.

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 10]
```diff
@@ -431,72 +440,52 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
-      dict({
-        'method': 'POST',
-      }),
-      reactionValue
-    ).catch((error) => {
-      dev().error(TAG, error);
+    return this.executeReactionRequest_('GET').then((response) => {
+      this.handleSuccessfulDataRetrieval_(response);
     });
   }
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction.js:446]: Removed since it was just calling executeReactionRequest("POST"), so I replaced it

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 11]
```diff
@@ -431,72 +440,49 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
     }
 
-    if (this.reactionId_ === null) {
+    if (!this.reactionId_) {
       const pageId = closest(dev().assertElement(this.element), (el) => {
         return el.tagName.toLowerCase() === 'amp-story-page';
       }).getAttribute('id');
-
-      this.reactionId_ = `CANONICAL_URL#page=${pageId}`;
+      this.reactionId_ = `CANONICAL_URL+${pageId}`;
```
Comment (mszylkowski) [inline extensions/amp-story/1.0/amp-story-reaction.js:466]: The #page=... was interfering with the fragment and braking the REST API so this <url>+<page> was introduced

File: src/url.js (hunk 6 above)
Comment (gmajoulet) [reply src/url.js:668]: Can we try to re-use browser logic instead?

We could do something like this directly from our Story codebase:
```
const anchor = document.createElement('a');
anchor.href = this.element.getAttribute('endpoint');
anchor.pathname += '/react';
endpoint = anchor.href;
```

File: extensions/amp-story/1.0/amp-story-reaction-quiz.js [hunk 12]
```diff
@@ -199,12 +200,12 @@ export class AmpStoryReactionQuiz extends AmpStoryReaction {
 
     const percentages = this.preprocessPercentages_(responseData);
 
-    toArray(responseData['responses']).forEach((response) => {
+    toArray(responseData).forEach((response, index) => {
```
[the remark this answers is not in the document]
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-reaction-quiz.js:203]: Changed the variable names to optionsData when used

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 13]
```diff
@@ -520,47 +504,37 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
   updateReactionOnDataRetrieval_(data) {
-    let selectedOptionKey;
-    /** @type {!Array} */ (data['responses']).forEach((response) => {
-      if (response.selectedByUser) {
-        selectedOptionKey = response.reactionValue;
-      }
-    });
-
-    if (selectedOptionKey === undefined) {
-      dev().error(TAG, `The user-selected reaction could not be found`);
```
[the remark this answers is not in the document]
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-reaction.js:553]: I think we shouldn't ever see this error, since this method assumed hasUserSelection=true but the backend doesn't have that field any more. The error essentially was catching the hasUserSelection=true and no option to be selected, which won't happen with the new API design.

File: EXTENSIONS_CSS_MAP [hunk 14]
```diff
@@ -0,0 +1 @@
...ority":"high"},"amp-video-0.1":{"name":"amp-video","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-video-docking-0.1":{"name":"amp-video-docking","version":"0.1","latestVersion":"0.1","hasCss":true},"amp-video-iframe-0.1":{"name":"amp-video-iframe","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-viqeo-player-0.1":{"name":"amp-viqeo-player","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-vk-0.1":{"name":"amp-vk","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-yotpo-0.1":{"name":"amp-yotpo","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-youtube-0.1":{"name":"amp-youtube","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-mowplayer-0.1":{"name":"amp-mowplayer","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-powr-player-0.1":{"name":"amp-powr-player","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-mraid-0.1":{"name":"amp-mraid","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-link-rewriter-0.1":{"name":"amp-link-rewriter","version":"0.1","latestVersion":"0.1","hasCss":false},"amp-minute-media-player-0.1":{"name":"amp-minute-media-player","version":"0.1","latestVersion":"0.1","hasCss":false}}
```
Comment (gmajoulet) [inline EXTENSIONS_CSS_MAP:1]: Please remove

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 15]
```diff
@@ -520,47 +505,45 @@ export class AmpStoryReaction extends AMP.BaseElement {
    * @private
    */
   handleSuccessfulDataRetrieval_(response) {
-    if (!(response && 'data' in response)) {
+    if (response && response['óptions']) {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-reaction.js:508]: s/`ó`/`o`

File: src/url.js [hunk 16]
```diff
@@ -650,3 +650,18 @@ export function checkCorsUrl(url) {
...
+
+/**
+ * Adds the path to the given url.
+ * @param {string} url
+ * @param {string} path
+ * @return {string}
+ */
+export function appendPathToUrl(url, path) {
+  if (!a) {
+    a = /** @type {!HTMLAnchorElement} */ (self.document.createElement('a'));
```
Comment (gmajoulet) [inline src/url.js:662]: @choumx knows better, but I think you might have to accept the anchor as a parameter as you can't use `self`.

**Comment by gmajoulet** [comment]
+ @choumx for `src/url.js` changes

File: src/url.js (hunk 16 above)
Comment (choumx) [reply src/url.js:662]: Yea, note that the only other usage is in a deprecated function. 

We also have a URL service for parsing URLs. See `url-impl.js` and `Services.urlForDoc`. You can use that for parsing a string into a URL object, and change this function to accept a URL object instead.

File: src/url.js [hunk 17]
```diff
@@ -650,3 +650,14 @@ export function checkCorsUrl(url) {
...
+/**
+ * Adds the path to the given url.
+ * @param {!Location} url
+ * @param {string} path
+ * @return {string}
+ */
+export function appendPathToUrl(url, path) {
+  const pathname = url.pathname + path.replace(/^\//, '');
+  return url.protocol + '//' + url.host + pathname + url.search + url.hash;
+}
```
Comment (mszylkowski) [inline src/url.js:663]: Changed to use url passed in

**Comment by mszylkowski** [comment]
ParseUrlWithA and url-impl were both restricted from being imported in amp-story-reaction so changed appendPathToUrl signature to provide the a tag to parse the url from.

File: extensions/amp-story/1.0/amp-story-reaction.js [hunk 18]
```diff
@@ -129,18 +132,43 @@ export class AmpStoryReaction extends AMP.BaseElement {
...
+    return this.rootEl_;
+  }
+
+  /**
+   * Gets the options.
+   * @visibleForTesting
+   * @return {!Array<!Element>}
+   */
+  getOptionElements() {
+    return toArray(
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-reaction.js:150]: Why did you remove the property that was caching these?

Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-reaction.js:150]: That was the original PR, but I'll add it here since it's not a big change

File: src/url.js (hunk 16 above)
Comment (mszylkowski) [reply src/url.js]: Those approaches required me to whitelist the reactions component to use url-impl or parseUrlWithA, but they were all essentially just getting an a tag and setting the href, so I changed the function to pass in the a tag directly to appendPathToUrl, lmk if this would work

File: src/url.js [hunk 19]
```diff
@@ -650,3 +650,17 @@ export function checkCorsUrl(url) {
...
+
+/**
+ * Adds the path to the given url.
+ *
+ * @param {!HTMLAnchorElement} a
+ * @param {string} url
+ * @param {string} path
+ * @return {string}
+ */
+export function appendPathToUrlWithA(a, url, path) {
```
Comment (choumx) [inline src/url.js:662]: Why can't we use the URL service here instead of creating an anchor and passing it in? The point of the URL service is to avoid creating unnecessary `<a>` elements and have caching of URLs.

File: src/url.js [hunk 20]
```diff
@@ -650,3 +650,17 @@ export function checkCorsUrl(url) {
...
+  a.href = url;
+  const pathname = a.pathname.replace(/\/?$/, '/') + path.replace(/^\//, '');
+  return a.protocol + '//' + a.host + pathname + a.search + a.hash;
```
Comment (choumx) [inline src/url.js:665]: Can we replace `a.protocol + '//' + a.host` with `a.origin`?

File: src/url.js [hunk 21]
```diff
@@ -650,3 +650,17 @@ export function checkCorsUrl(url) {
...
+ * Adds the path to the given url.
+ *
+ * @param {!HTMLAnchorElement} a
+ * @param {string} url
+ * @param {string} path
+ * @return {string}
+ */
+export function appendPathToUrlWithA(a, url, path) {
+  a.href = url;
+  const pathname = a.pathname.replace(/\/?$/, '/') + path.replace(/^\//, '');
```
Comment (choumx) [inline src/url.js:664]: When will `pathname` have a trailing `?`?

File: src/url.js (hunk 16 above)
Comment (choumx) [reply src/url.js]: `Services.urlForDoc` in `services.js` is the right way to get a service object.

`parseUrlWithA` is restricted precisely to avoid callers from creating their own anchor elements. :)

Comment (mszylkowski) [reply src/url.js]: Thanks for the more detailed explanation, fixed it on upcoming commit

File: src/url.js (hunk 19 above)
Comment (mszylkowski) [reply src/url.js:662]: Done!

File: src/url.js (hunk 20 above)
Comment (mszylkowski) [reply src/url.js:665]: Yes, done!

File: src/url.js (hunk 21 above)
Comment (mszylkowski) [reply src/url.js:664]: \/?$ detects "zero or one slashes at the end", the question mark is being used as the regex quantifier for the escaped forward slash. This replace ensures it ends with one forward slash ("/hello" -> "/hello/") so that the concatenation contains a slash in the middle.
